src/tools/perf/perftest.c

```diff
@@ -970,10 +991,66 @@ static unsigned mpi_rte_group_index(void *rte_group)
     return rank;
 }
 
-static void mpi_rte_barrier(void *rte_group)
+static void mpi_rte_barrier(void *rte_group, void (*progress)(void *arg),
+                            void *arg)
 {
+    int group_size, my_rank, i;
+    MPI_Request *reqs;
+    int nreqs = 0;
+    int dummy;
+    int flag;
+
 #pragma omp master
-    MPI_Barrier(MPI_COMM_WORLD);
+
+    /*
+     * Naive barrier implementation over send/recv, to call user progress
+     * while waiting for barrier to complete.
+     */
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
```

Comment:
what about use non-blocking MPI barrier?
https://www.mpich.org/static/docs/v3.2/www3/MPI_Ibarrier.html

Reply:
surely would be simpler, but didn't want to depend on MPI-3..